src/main/java/org/ensime/maven/plugins/ensime/ServerIndexMojo.java

```diff
@@ -0,0 +1,145 @@
+    }
+    List<File> humanJars = partitioned.get(false);
+    if (humanJars != null) {
+      orderedJars.addAll(humanJars);
+    }
+    return orderedJars;
+  }
+
+  private static void launchEnsimeServer(
+      String javaCommand,
+      List<File> orderedClasspath,
+      List<String> jvmFlags) throws IOException, InterruptedException {
+    ProcessBuilder builder = new ProcessBuilder();
+    List<String> fullCommand = new ArrayList<>();
+    fullCommand.add(javaCommand);
+    // JVM flags
+    fullCommand.addAll(jvmFlags);
+    // Class path
+    fullCommand.add("-cp");
+    fullCommand.add(orderedClasspath
+        .stream()
+        .map(File::getPath)
+        .collect(Collectors.joining(":"))
+    );
+    // Main class
+    fullCommand.add("org.ensime.server.Server");
+    // Start the server and wait for it to finish.
+    builder.inheritIO()
+        .command(fullCommand)
+        .start()
+        .waitFor();
+  }
+
+  @Override
+  public void execute() {
+    // Preferably we would read the .ensime config but that would mean we would either write our own
+    // S-Expression parser or we would have to use a library to do it. Instead, we will regenerate
+    // the maven project metadata and an EnsimeConfig object from it, which will then be used for
+    // starting the ensime server for indexing.
+
+    Properties properties = new Properties();
+    EnsimeConfigGenerator generator = new EnsimeConfigGenerator(project,
+        repoSystem, session, properties, ensimeServerVersion, ensimeScalaVersion,
+        getLog());
+    EnsimeConfig ensimeConfig = generator.generateConfig();
+
+    String javaCommand = String.format("%s/bin/java", ensimeConfig.getJavaHome());
+    List<File> classPathJars = new ArrayList<>();
+    classPathJars.addAll(ensimeConfig.getEnsimeServerJars());
+    classPathJars.addAll(ensimeConfig.getScalaCompilerJars());
+    classPathJars.add(new File(String.format("%s/lib/tools.jar", ensimeConfig.getJavaHome())));
+    List<File> orderedClasspath = orderJars(classPathJars);
+    List<String> jvmFlags = new ArrayList<>();
+    jvmFlags.add("-Xms4g");
+    jvmFlags.add("-Xmx4g");
+    jvmFlags.add("-XX:StringTableSize=1000003");
+    jvmFlags.add("-XX:+UnlockExperimentalVMOptions");
+    jvmFlags.add("-XX:SymbolTableSize=1000003");
+    jvmFlags.add("-Densime.config=.ensime");
+    jvmFlags.add("-Densime.exitAfterIndex=true");
+    jvmFlags.addAll(ensimeConfig.getJavaFlags());
+    File cacheDir = ensimeConfig.getCacheDir();
+    if (!cacheDir.exists() && !cacheDir.mkdirs()) {
+      throw new RuntimeException(
+          "Unable to create ensime cache directory. Please check permissions.");
+    }
+    try {
+      launchEnsimeServer(javaCommand, orderedClasspath, jvmFlags);
+    } catch (IOException e) {
+      getLog().error("Unable to start ensime server for indexing.", e);
+    } catch (InterruptedException e) {
+      getLog().error("Interrupted while indexing.", e);
+    }
+  }
```

Comment:
I really don't like this hidden dependency. Adding a new target or renaming one would potentially break this target.

Including a library or even writing a simple parser would have been a safer way to go

Reply:
@amanjpro Which dependency are you referring to? The dependency between the generate goal and the server index goal?

Reply:
Yes

Reply:
@amanjpro feel free to revert my version bump, if any changes are made. I felt it was okay since we are doing it in sbt plugin. But I agree removing that dependency would be good.

Reply:
@amanjpro Forgive me, but I don't understand what you mean by:

> Adding a new target or renaming one would potentially break this target.

The `serverIndex` goal right now has no dependency on the `generate` goal. It simply uses the EnsimeConfigGenerator class as a helper to generate the maven metadata before launching an ensime server process. I would argue that parsing the ensime config will _add_ a dependency between the two goals because it would force `serverIndex` to run after `generate` is complete. The way it is right now, this is not the case - you can run `serverIndex` without running `generate`.

@yyadavalli Am I missing something? Thank you for any further elaboration.

Reply:
I think reusing `EnsimeConfigGenerator` is a dependency because if for some reason the generator is not working I should still be able to use serverIndex. It is possible to hand write or generate the `.ensime` with other build tools. 

It is kind of a chicken and egg problem, so I am okay either way. But decoupling `serverIndex` from other parts of the codebase and depending on a valid `.ensime` file is slightly better.

Reply:
@yyadavalli thank you for the clarification. I understand what you're saying. 

But if `generate` is broken, we're unable to perform the main task that this plugin was built for and instead we have to rely on another build tool. That seems like a much bigger problem than having this dependency between goals. To me, this doesn't quite seem like a 'hidden' dependency, but rather reusing code between two tasks that are closely interrelated. Maybe the implementation of this reuse is flawed and I can certainly abide by refactoring suggestions, but I think reusing EnsimeConfigGenerator to retrieve common metadata about the project is not a bad thing because the metadata is useful for many different goals in the plugin. This kind of code sharing is very common in other maven plugins because plugin goals generally cohere and require similar inputs, and to avoid code duplication, it makes sense to use a utility or a common class.

If @amanjpro insists on adding a new library or a parser just to read the .ensime file I will do so but could we possibly add that as a new issue? I am more than happy to work on that as well.

Reply:
I thought about it again, and if users want to customize serverIndex as you mentioned, the plugin has to read the .ensime file. I think that is a strong reason and I'm convinced. Can I create a new pull request?

Reply:
Go ahead create the PR

Reply:
Ah, sorry I am not complaining. I am grateful for the PR and the PR is already merged. My worry is just a mere fear that we might end-up having issues in the future. A better solution can definitely wait.

I won't have time to make a release soon, if somebody is willing to do it, he is definitely welcome to do it.

Just a note, adding an issue regarding this hidden dependency might be a good first step to start.

As you proposed good way to fix this, would be to have a config generator in a utility class that takes the needed parameters. That way the compiler would better protect us when a change is made in the future.

Reply:
Cool, let's take this up in a new PR. @amanjpro can you please publish the 1.4.0 release to maven. I'm having trouble publishing from my end.

Reply:
@amanjpro Absolutely, I know you're not complaining. :) I was trying to understand your point about the hidden dependency. In my humble opinion, I think having the dependency between generate and serverIndex isn't a deal breaker, but we should ideally read the .ensime file because there may be customizations like java-flags or environment variables that are not picked up just from the maven project. I think this would ease testing and so I think it's worth doing in a separate PR. 

As always, thank you all for any and all feedback!

Reply:
Here's the new issue to read in the .ensime file: https://github.com/ensime/ensime-maven/issues/62.